libvuln/updates/manager.go

```diff
@@ -0,0 +1,256 @@
+			if err := us.Add(u); err != nil {
+				log.Warn().Err(err).Str("updater", u.Name()).Msg("Duplicate updater, skipping")
+			}
+		}
+		m.factories["outOfTree"] = driver.StaticSet(us)
+	}
+
+	// these options can be ran order independent.
+	for _, opt := range opts {
+		opt(m)
+	}
+
+	return m, nil
+}
+
+// Start will run updaters at the given interval.
+//
+// Start is designed to be ran as a go routine.
+// Cancel the provided ctx to end the updater loop.
+func (m *Manager) Start(ctx context.Context) error {
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "libvuln/updates/Manager.Start").
+		Logger()
+	ctx = log.WithContext(ctx)
+	if m.interval == 0 {
+		return fmt.Errorf("Manager must be configured with an interval to start")
+	}
+	log.Info().Str("interval", m.interval.String()).Msg("Starting background updaters")
+
+	t := time.NewTicker(m.interval)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-t.C:
+			m.Run(ctx)
+		}
+	}
+}
+
+// Run constructs updaters from factories, configures them
+// and runs them in Manager.batchSize batches.
+//
+// Run is safe to call at anytime, regardless of whether
+// background updaters are running.
+func (m *Manager) Run(ctx context.Context) error {
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "libvuln/updates/Manager.Run").
+		Logger()
+	ctx = log.WithContext(ctx)
+
+	updaters := []driver.Updater{}
+	// constructing updater sets may require network access,
+	// depending on the factory.
+	// if construction fails we will simply ignore those updater
+	// sets.
+	for _, factory := range m.factories {
+		set, err := factory.UpdaterSet(ctx)
+		if err != nil {
+			log.Error().Err(err).Msg("Error while calling factory for updater set. Updaters in this factory will be excluded from batch.")
+			continue
+		}
+		updaters = append(updaters, set.Updaters()...)
+	}
+
+	// reconfigure updaters
+	for _, u := range updaters {
+		f, fOK := u.(driver.Configurable)
+		cfg, cfgOK := m.configs[u.Name()]
+		if fOK && cfgOK {
+			if err := f.Configure(ctx, cfg, nil); err != nil {
+				log.Warn().Err(err).Str("updater", u.Name()).Msg("Failed configuring updaters. Updater will be excluded from batch.")
+				continue
+			}
+		}
+	}
+
+	log.Info().Int("total", len(updaters)).Int("batchSize", m.batchSize).Msg("Batching updaters...")
+
+	sem := semaphore.NewWeighted(int64(m.batchSize))
+	errChan := make(chan error, len(updaters)+1) // +1 for a potential ctx error
+	for i := range updaters {
+
+		err := sem.Acquire(ctx, 1)
+		if err != nil {
+			return fmt.Errorf("Failed to acquire semaphore")
+		}
+
+		go func(u driver.Updater) {
+			defer sem.Release(1)
+
+			if m.pool != nil {
+				lock := postgres.NewPool(m.pool, 0)
+				ok, err := lock.TryLock(ctx, u.Name())
+				if err != nil {
+					errChan <- err
+					return
+				}
+				if !ok {
+					log.Debug().Str("updater", u.Name()).Msg("Lock acquisition failed. Another process is running this updater.")
```

Comment:
I don't think a debug log for a common case needs to be so verbose.

```suggestion
					log.Debug().Str("updater", u.Name()).Msg("lock acquisition failed")
```

Reply:
As a debugger, that doesn't provide me with much context to why the lock failed. Most ops engineers are going to look at our debug code and not have the context we do as the authors.

Reply:
updated to 
```
log.Debug().Str("updater", u.Name()).Msg("another process running updater, excluding from run.")
```